KAFKA-13785: [6/N][Emit final] emit final for TimeWindowedKStreamImpl

lihaosky opened the pull request:
### Description
Initial implementation to emit final for TimeWindowedKStreamImpl. This PR is on top of https://github.com/apache/kafka/pull/12030

### Testing
Unit test and integration test

lihaosky committed:
[Emit final][5/N] emit final for TimeWindowedKStreamImpl

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:
```diff
@@ -184,6 +203,52 @@ public void process(final Record<KIn, VIn> record) {
+                    return;
+                }
+
+                // Emit final deosn't make sense for some window like unlimitedWindow
```
typo

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimeWindowedKStreamImpl.java:
```diff
@@ -224,7 +225,15 @@
-            supplier = Stores.persistentTimestampedWindowStore(
+            // TODO: use different store name for emit final case?
+            supplier = emitFinal ?
+                Stores.persistentTimeOrderedWindowStore(
```
Should this be updated?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimeWindowedKStreamImpl.java:
```diff
@@ -224,7 +225,15 @@
+                materialized.storeName(),
+                Duration.ofMillis(retentionPeriod),
+                Duration.ofMillis(windows.size()),
+                false, false)
```
Should `hasIndex` be true here?

lihaosky replied:
Seems `windowStore` in window aggregate is only used by `windowStore.fetch(key, window.start());`, so I don't need to use index as fetching from time ordered schema is also a point get.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/Stores.java:
```diff
@@ -160,6 +161,35 @@ public String metricsScope() {
         };
     }
 
+    /**
```
This class's changes should be reverted.

lihaosky replied:
Yes. It was previously on topic another change

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:
```diff
@@ -184,6 +203,52 @@ public void process(final Record<KIn, VIn> record) {
+
+                // Emit final deosn't make sense for some window like unlimitedWindow
+                // This should be checked in upstream before pass emit final in or checked here
+                final long closeWindowStart = closeTime - windows.size();
```
nit: would it be better named `emitWindowStart`?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:
```diff
@@ -184,6 +203,52 @@ public void process(final Record<KIn, VIn> record) {
+                    KeyValue<Windowed<KIn>, ValueAndTimestamp<VAgg>> kv = windowToEmit.next();
+                    tupleForwarder.maybeForward(
+                        record.withKey(kv.key)
+                            .withValue(new Change<>(kv.value.value(), null))
```
This is not a suggestion: we are sending `null` with the guarantee that we should have never forward for this key before. I think a good test case coverage would be to have a windowed aggregation emit final, followed by a join. The join results would need both the old/new values to be able to correct, and if emit final we should only emit once, with old value setting to null.

guozhangwang reviewed (commented):
Made a pass; we need to rebase this branch a bit and after that I will review again.

lihaosky reviewed (commented):
@guozhangwang . Yeah. I'll need to rebase and add unit tests. Merging this needs https://github.com/apache/kafka/pull/11829 which is also in review. I'll rebase this on top of https://github.com/apache/kafka/pull/11829 and add unit test. I can start perf testing first before this can be merged.

lihaosky committed:
address comments and bug fix

lihaosky committed:
unit tests

lihaosky committed:
integration test

lihaosky commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimeWindowedKStreamImpl.java:
```diff
@@ -232,11 +247,19 @@
-                        false
+                    supplier = emitStrategy.type() == StrategyType.ON_WINDOW_CLOSE ?
+                        RocksDbIndexedTimeOrderedWindowBytesStoreSupplier.create(
+                            materialized.storeName(),
```
@guozhangwang , should we call this another name to avoid overriding existing emit eager's store?

mjsax replied:
For what scenario should it happen that we have two stores with the same name? (I believe there is no problem and the code is fine, but maybe I miss something?)

lihaosky replied:
If a user just enables `emit final` for their existing topology which uses `emit eager`, will it cause it to use existing state store which has wrong data format etc?

mjsax replied:
I think it's ok to declare this a non-supported pattern? We should call it out in the docs.

guozhangwang replied:
Sorry I got to this late --- also sgtm.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/EmitStrategy.java:
```diff
@@ -0,0 +1,65 @@
+
+    /**
+     * This strategy indicates that the aggregated result for a window will only be outputted when the
+     * window closes instead of when there's an update to the window.
```
Can we add details what "closes" means, ie, _when_ a window is closed?

Should we also mention caching? Or would this go into the weeds?

lihaosky replied:
Sure. I feel this is not related to caching.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/EmitStrategy.java:
```diff
@@ -0,0 +1,65 @@
+     * window closes instead of when there's an update to the window.
+     *
+     * <p>This strategy should only be used for window which can close. For example, it doesn't make sense
+     * to be used with {@link UnlimitedWindow}
```
What happens if it's used anyway?

(nit: missing `.` at the end of the sentence)

lihaosky replied:
Will update with details.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:
```diff
@@ -80,22 +109,54 @@ public void enableSendingOldValues() {
+                    sendOldValues);
+            }
+
+            log.info("EmitStrategy=" + emitStrategy.type());
```
Might debug level be better?

lihaosky replied:
I guess this won't be called many times and flood the log while this is helpful information. Will this be printed somewhere else (i.e. topology)?

mjsax replied:
> while this is helpful information

Well, is it? To me, it sounds like as if a filter logs: "you are using a filter". -- The concern is not about spamming the logs, but it seems not to be helpful but just noise in the logs. -- I am also ok to leave it in.

lihaosky replied:
Is filter logged in the topology? This isn't logged in topology though :)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:
```diff
@@ -80,22 +109,54 @@ public void enableSendingOldValues() {
+                    1000L
+                );
+                timeTracker.setEmitInterval(emitInterval);
+                log.info("EmitInterval=" + emitInterval);
```
maybe better to use debug level?

lihaosky replied:
Similar to previous comment, I feel this won't be called many time (just once when creating the processor?). And the config won't be printed out when we print all configs since this is an internal config. This will be convenient for debugging.

mjsax replied:
It's not about spamming but it's a question if it's useful? Also seems to leak an internal config that we might want to keep hidden?

lihaosky replied:
I put it here since we don't log it since it's internal config... I'm ok to delete this though

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:
```diff
@@ -184,6 +247,75 @@ public void process(final Record<KIn, VIn> record) {
+
+            // Schedule next emit time based on now to avoid the case that if system time jumps a lot,
+            // this can be triggered everytime
+            timeTracker.nextTimeToEmit = now;
```
Did we do the same thing for stream-stream join? I don't think we did? Might be worth to do there, too? (Not part of this PR...)

lihaosky replied:
stream-stream join didn't do this. Created https://issues.apache.org/jira/browse/KAFKA-13817 to track

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimeWindowedKStreamImpl.java:
```diff
@@ -202,12 +207,22 @@
                 materializedInternal.valueSerde());
     }
 
+    //@Override
```
Should not be a comment?

lihaosky replied:
This method is not added to the interface `TimeWindowedKStream` yet since `SlidingWindowedKStreamImpl` also implements the interface. I plan to add it to interface together with `SlidingWindowedKStreamImpl` implementation.

lihaosky committed:
cache for time ordered window store

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:
```diff
@@ -80,22 +109,54 @@ public void enableSendingOldValues() {
             final StreamsMetricsImpl metrics = internalProcessorContext.metrics();
             final String threadId = Thread.currentThread().getName();
             droppedRecordsSensor = droppedRecordsSensor(threadId, context.taskId().toString(), metrics);
+            emittedRecordsSensor = emittedRecordsSensor(threadId, context.taskId().toString(), metrics);
```
While useful, I cannot remember that we discussed to add metrics on the KIP. Metrics are public API and should be included in the KIP.

lihaosky replied:
I'm ok removing this if the process involves another round of discussion and voting 😃

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:
```diff
@@ -80,22 +109,54 @@ public void enableSendingOldValues() {
+        private long lastEmitCloseTime = ConsumerRecord.NO_TIMESTAMP;
+        private InternalProcessorContext<Windowed<KIn>, Change<VAgg>> internalProcessorContext;
+        private final TimeTracker timeTracker = new TimeTracker();
+        private final Time time = Time.SYSTEM;
```
We should use `Time.System` but pass in a `Time` object from outside -- otherwise, we cannot mock it.

lihaosky replied:
Read code again, this is only used in measure latency, I don't see we tested measuring latency for `Time`. Can we add it to constructor when we need to mock it?

mjsax replied:
If we don't test is right now, ok to leave as is. It's just a general rule that we should "never" use `Time.SYSTEM` but always pass in a `Time` object to write future proof/mockable code from the beginning.

If there are other cases like this, might be worth to clean up as follow up. Maybe we should file some follow up tickets?

lihaosky replied:
Sure. Created https://issues.apache.org/jira/browse/KAFKA-13824

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:
```diff
@@ -184,6 +247,75 @@ public void process(final Record<KIn, VIn> record) {
+            timeTracker.advanceNextTimeToEmit();
+
+            // Close time does not progress
+            if (lastEmitCloseTime != ConsumerRecord.NO_TIMESTAMP && lastEmitCloseTime >= closeTime) {
```
Would `lastEmitCloseTime` not always be `<= closeTime` ?

lihaosky replied:
If we have multiple source topics in a task and `lastEmitCloseTime` is committed successfully for some sources and commit offset is not committed successfully for some sources, is it possible we restored `lastEmitCloseTime` and start processing from earlier time? In this case, `lastEmitCloseTime` can be larger than `closeTime`? Maybe log warning if `lastEmitCloseTime` is larger?

mjsax replied:
Maybe. Sounds like a weird corner case that could lead to duplicate output (what is ok for ALOS)? -- Might also be ok to just leave it in -- seems there is no concern about correctness?

lihaosky replied:
Yeah. No concern of correctness.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:
```diff
@@ -184,6 +247,75 @@ public void process(final Record<KIn, VIn> record) {
+                return;
+            }
+
+            // Because we only get here when emitWindowStart > 0 which means closeTime > windows.size()
```
nit: `emitWindowStart >= 0`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:
```diff
@@ -184,6 +247,75 @@ public void process(final Record<KIn, VIn> record) {
+
+            // Because we only get here when emitWindowStart > 0 which means closeTime > windows.size()
+            // Since we set lastEmitCloseTime to closeTime before storing to processor metadata
+            // lastEmitCloseTime - windows.size() is always > 0
```
The comment make sense, but I am not sure what the point is you try to make?

lihaosky replied:
Trying to say `lastEmitCloseTime - windows.size()` won't be less than or equal to 0 and the reason

guozhangwang replied:
I will rephrase the comment a bit to make it clearer.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:
```diff
@@ -184,6 +247,75 @@ public void process(final Record<KIn, VIn> record) {
+                // Don't fetch store if the new emit window close time doesn't progress enough to cover next
+                // window
+                if (matchedCloseWindows.equals(matchedEmitWindows)) {
+                    log.debug("no new windows to emit. LastEmitCloseTime={}, newCloseTime={}",
```
Better log this at `trace` level?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:
```diff
@@ -184,6 +247,75 @@ public void process(final Record<KIn, VIn> record) {
+
+                // Don't fetch store if the new emit window close time doesn't progress enough to cover next
+                // window
+                if (matchedCloseWindows.equals(matchedEmitWindows)) {
```
Is this the same as `emitWindowStart ==  lastEmitWindowStart`? If yes, comparing two longs seems to be more efficient than comparing to `Maps`

mjsax replied:
As we `.fetchAll(lastEmitWindowStart + 1, emitWindowStart);` and both upper and lower bound are inclusive, would we not need that `emitWindowStart >= lastEmitWindowStart+1` to have at least one window that we could potentially emit? Otherwise, lower bound is larger than upper bound and the time-range is not well defined?

lihaosky replied:
> Is this the same as emitWindowStart == lastEmitWindowStart

Nope. They may not be the same. `emitWindowStart` is based on current `closeTime` and `lastEmitWindowStart` is based last emitted close time. These are not the same. The logic here to detect if `closeTime` doesn't progress much to emit new windows.

For example, for tumbling window length 10sec and close time 11, window [0, 10) will be emitted, if close time progress to 12, window to emit is still [0, 10).

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:
```diff
@@ -184,6 +247,75 @@ public void process(final Record<KIn, VIn> record) {
+                    record.withKey(kv.key)
+                        .withValue(new Change<>(kv.value.value(), null))
+                        .withTimestamp(kv.value.timestamp())
+                        .withHeaders(null)); // Don't set header
```
Based on what we do elsewhere, I think we should set the header as `context.header()` -- I understand that it's semantically questionable, but we should be consistent with what we do elsewhere (and maybe clean it up in a consistent manner via a KIP at some point).

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:
```diff
@@ -184,6 +247,75 @@ public void process(final Record<KIn, VIn> record) {
+                return;
+            }
+
+            final long emitWindowStart = closeTime - windows.size();
```
It took me some time to figure out what this variable really is. It's the upper bound of the emit range (based on window-start) right?

Maybe we can find a better name? Maybe `emitRangeUpperBoundInclusive` ?

Similarly it might be helpful to introduce an `emitRangeLowerBoundInclusive` variable?

lihaosky replied:
Yeah. Will rename. Naming is hard...

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:
```diff
@@ -184,6 +247,75 @@ public void process(final Record<KIn, VIn> record) {
+            emittedRecordsSensor.record(emittedCount);
+
+            lastEmitCloseTime = closeTime;
+            internalProcessorContext.addProcessorMetadataKeyValue(storeName, closeTime);
```
After considering the logic, I am wondering if it might be simpler to not store the latest "closeTime" but the latest window-start time for which we did emit (ie, `emitRangeLowerBoundInclusive`)? It's of course just shifted by window-size and grace-period but might make the code easier to structure and read/understand?

lihaosky replied:
I thought about this and thought it might be more consistent if we shift both `lastEmitCloseTime` and `closeTime`. If we just store `lastEmitStartTime`, we only shift `closeTime` when processing. Also I feel it's easier to store `closeTime` as it's easier to see if a window closes without shifting it and see.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/TimeWindowedKStreamIntegrationTest.java:
```diff
@@ -0,0 +1,497 @@
+public class TimeWindowedKStreamIntegrationTest {
+    private static final int NUM_BROKERS = 1;
+
+    public static final EmbeddedKafkaCluster CLUSTER = new EmbeddedKafkaCluster(NUM_BROKERS,
```
Why do we test using `EmbeddedKafkaCluster`? Seems using `TopologyTestDriver` should be sufficient and would be less heavy?

lihaosky replied:
I saw `EmbeddedKafkaCluster` is always used for Integration testing and `TopologyTestDriver` is used for unit test?

mjsax replied:
Guess, it depends on your definition of "unit" and "integration" tests... :)

To me, using `EmbeddedKafkaCluster` is necessary to test something we can not test with TTD -- in the end we want to test "business logic" ie, the "emit final" implementation -- what do we gain (ie, test better) using `EmbeddedKafkaCluster` compared to TTD?

lihaosky replied:
Because `EmbeddedKafkaCluster` is real environment while TTD is mocking environment, this is meant to cover any gaps TTD has from `EmbeddedKafkaCluster` I think. Also I'm not sure TTD can test something like "restoration from changelog" easily. BTW, @vvcephei mentioned we should have tests for both TTD and `EmbeddedKafkaCluster` unless I understand the context wrong...

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/TimeWindowedKStreamIntegrationTest.java:
```diff
@@ -0,0 +1,497 @@
+            new KeyValueTimestamp<>("A", "1", 5),
+            new KeyValueTimestamp<>("A", "1", 10), // close [0, 10)
+            new KeyValueTimestamp<>("B", "2", 6),  // late and skip
+            new KeyValueTimestamp<>("B", "2", 11), // close [0, 10)
```
Why would this close `[0,10)` ? It was already closed?

lihaosky replied:
Yeah. It's closed. Will update comment

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/TimeWindowedKStreamIntegrationTest.java:
```diff
@@ -0,0 +1,497 @@
+            new KeyValueTimestamp<>("A", "1", 0),
+            new KeyValueTimestamp<>("A", "1", 5),
+            new KeyValueTimestamp<>("A", "1", 10), // close [0, 10)
+            new KeyValueTimestamp<>("B", "2", 6),  // late and skip
```
Why would this record be skipped? You define a hopping window, and while window `[0,10)` is closed, window `[5,15)` is still open and the record should still go into this second window?

lihaosky replied:
Yeah. It does go to second window. Will update the comment

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/TimeWindowedKStreamIntegrationTest.java:
```diff
@@ -0,0 +1,497 @@
+        produceMessages(
+            streamOneInput,
+            new KeyValueTimestamp<>("A", "1", 0),
+            new KeyValueTimestamp<>("A", "1", 5),
```
The output might be easier to read/understand if we use unique values, `1` to `7` for the seven input records.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/TimeWindowedKStreamImplTest.java:
```diff
@@ -42,84 +47,142 @@
+                asList(
+                    new KeyValueTimestamp<>(KEY_1_WINDOW_0, 2L, 15L),
+                    new KeyValueTimestamp<>(KEY_1_WINDOW_1, 1L, 500L),
+                    new KeyValueTimestamp<>(KEY_2_WINDOW_1, 2L, 550L)
```
Why does the order change?

lihaosky replied:
Didn't get your question. The fetch for emit final is based on `windowStart` order.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/TimeWindowedKStreamImplTest.java:
```diff
@@ -158,21 +235,23 @@ public void shouldMaterializeCount() {
                     KeyValue.pair(new Windowed<>("1", new TimeWindow(500, 1000)), 1L),
-                    KeyValue.pair(new Windowed<>("2", new TimeWindow(500, 1000)), 2L))));
+                    KeyValue.pair(new Windowed<>("2", new TimeWindow(500, 1000)), 2L),
+                    KeyValue.pair(new Windowed<>("2", new TimeWindow(1000, 1500)), 1L))));
```
Why do we get one more output record?

lihaosky replied:
I added one more input in `processData`: `inputTopic.pipeInput("2", "30", 1000L);`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/TimeWindowedKStreamIntegrationTest.java:
```diff
@@ -0,0 +1,497 @@
+            streamOneInput,
+            new KeyValueTimestamp<>("A", "1", 0),
+            new KeyValueTimestamp<>("A", "1", 5),
+            new KeyValueTimestamp<>("A", "1", 10), // close [-5, 5)
```
Does `[-5,5)` exists?

lihaosky replied:
Not really 🥲 . Will update

lihaosky committed:
comments

lihaosky committed:
checkstyle

lihaosky committed:
comments

lihaosky committed:
update metrics

lihaosky committed:
checkstyle

lihaosky committed:
checkstyle again...

lihaosky committed:
delete log

lihaosky committed:
optimization

lihaosky committed:
checkstyle refactor

lihaosky committed:
update metrics name

cmccabe committed:
KAFKA-13807: Fix incrementalAlterConfig and refactor some things (#12033)

Ensure that we can set log.flush.interval.ms at the broker or cluster level via
IncrementalAlterConfigs. This was broken by KAFKA-13749, which added log.flush.interval.ms as the
second synonym rather than the first. Add a regression test to DynamicConfigChangeTest.

Create ControllerRequestContext and pass it to every controller API. This gives us a uniform way to
pass through information like the deadline (if there is one) and the Kafka principal which is
making the request (in the future we will want to log this information).

In ControllerApis, enforce a timeout for broker heartbeat requests which is equal to the heartbeat
request interval, to avoid heartbeats piling up on the controller queue. This should have been done
previously, but we overlooked it.

Add a builder for ClusterControlManager and ReplicationControlManager to avoid the need to deal
with a lot of churn (especially in test code) whenever a new constructor parameter gets added for
one of these.

In ControllerConfigurationValidator, create a separate function for when we just want to validate
that a ConfigResource is a valid target for DescribeConfigs. Previously we had been re-using the
validation code for IncrementalAlterConfigs, but this was messy.

Split out the replica placement code into a separate package and reorganize it a bit.

Reviewers: David Arthur <mumrah@gmail.com

RivenSun2 committed:
MINOR: Correct spelling errors in KafkaRaftClient (#12061)

Correct spelling errors in KafkaRaftClient

Reviewers: José Armando García Sancio <jsancio@users.noreply.github.com>

dajac committed:
MINOR: Fix TestDowngrade.test_upgrade_and_downgrade (#12027)

The second validation does not verify the second bounce because the verified producer and the verified consumer are stopped in `self.run_validation`. This means that the second `run_validation` just spit out the same information as the first one. Instead, we should just run the validation at the end.

Reviewers: Jason Gustafson <jason@confluent.io>

showuon committed:
KAFKA-8785: fix request timeout by waiting for metadata cache up-to-date (#11681)

The reason why this test is flaky is because we have race condition at the beginning of the test, when brokers are staring up, and the adminClient is requesting for brokers metadata. Once the adminClient only got partial metadata, the test will fail, because in these tests, brokers will be shutdown to test leader election.

Fix this issue by explicitly waiting for metadata cache up-to-date in waitForReadyBrokers, and let admin client get created after waitForReadyBrokers.

Reviewers: Jason Gustafson <jason@confluent.io>, David Jacot <djacot@confluent.io>, dengziming <dengziming1993@gmail.com>

dengziming committed:
KAFKA-13832: Fix flaky testAlterAssignment (#12060)

In KRaft mode the metadata is not propagate in time, so we should should wait for it before make assertions.

Reviewers:  Luke Chen <showuon@gmail.com>

jeqo committed:
KAFKA-13654: Extend KStream process with new Processor API (#11993)

Updates the KStream process API to cover the use cases
of both process and transform, and deprecate the KStream transform API.

Implements KIP-820

Reviewer: John Roesler <vvcephei@apache.org>

cmccabe committed:
KAFKA-13835: Fix two bugs related to dynamic broker configs in KRaft (#12063)

Fix two bugs related to dynamic broker configs in KRaft. The first bug is that we are calling reloadUpdatedFilesWithoutConfigChange when a topic configuration is changed, but not when a
broker configuration is changed. This is backwards. This function must be called only for broker 
configs, and never for topic configs or cluster configs.

The second bug is that there were several configurations such as max.connections which are related
to broker listeners, but which do not involve changing the registered listeners. We should support
these configurations in KRaft. This PR fixes the configuration change validation to support this case.

Reviewers: Jason Gustafson <jason@confluent.io>, Matthew de Detrich <mdedetrich@gmail.com>

divijvaidya committed:
KAFKA-10095: Add stricter assertion in LogCleanerManagerTest (#12004)


Reviewers: Mickael Maison <mickael.maison@gmail.com>

mimaison committed:
MINOR: Scala cleanups in core (#12058)


Reviewers: Luke Chen <showuon@gmail.com>, Divij Vaidya <diviv@amazon.com>

RivenSun2 committed:
KAFKA-13799: Improve documentation for Kafka zero-copy (#12052)

Improve documentation for Kafka zero-copy. Kafka combines pagecache and zero-copy to greatly improve message consumption efficiency. But zero-copy only works in PlaintextTransportLayer.

Reviewers: Divij Vaidya <divijvaidya13@gmail.com>, Guozhang Wang <wangguoz@gmail.com>

lihaosky committed:
KAFKA-13785: [5/N][emit final] cache for time ordered window store (#12030)

A new cache for RocksDBTimeOrderedWindowStore. Need this because RocksDBTimeOrderedWindowStore's key ordering is different from CachingWindowStore which has issues for MergedSortedCacheWindowStoreIterator

Reviewers: Matthias J. Sax <mjsax@apache.org>, Guozhang Wang <wangguoz@gmail.com>

sayantanu-dey committed:
KAFKA-13588: consolidate `changelogFor` methods to simplify the generation of internal topic names (#11703)

Reviewers: Guozhang Wang <wangguoz@gmail.com>

mjsax committed:
HOTFIX: fix broken trunk due to conflicting and overlapping commits (#12074)

Reviewers: Victoria Xia <victoria.xia@confluent.io>, David Arthur <mumrah@gmail.com>

RivenSun2 committed:
MINOR: cleanup for postProcessAndValidateIdempotenceConfigs method (#12069)

Reviewers: Luke Chen <showuon@gmail.com>

ruanliang-hualun committed:
KAFKA-13834: fix drain batch starving issue (#12066)

In drainBatchesForOneNode method, there's possibility causing some partitions in a node will never get picked. Fix this issue by maintaining a drainIndex for each node.

Reviewers: Luke Chen <showuon@gmail.com>, RivenSun <91005273+RivenSun2@users.noreply.github.com>

clayburn committed:
MINOR: Improve Gradle Caching and Fix Deprecations (#12003)

* Fix UP-TO-DATE check in `create*VersionFile` tasks

`create*VersionFile` tasks explicitly declared output UP-TO-DATE status
 as being false. This change properly sets the inputs to
`create*VersionFile` tasks to the `commitId` and `version` values and
sets `receiptFile` locally rather than in an extra property.

* Enable output caching for `process*Messages` tasks

`process*Messages` tasks did not have output caching enabled. This
change enables that caching, as well as setting a property name and
RELATIVE path sensitivity.

* Fix existing Gradle deprecations

Replaces `JavaExec#main` with `JavaExec#mainClass`

Replaces `Report#destination` with `Report#outputLocation`

Adds a `generator` configuration to projects that need to resolve
the `generator` project (rather than referencing the runtimeClasspath
of the `generator` project from other project contexts.

Reviewers: Mickael Maison <mickael.maison@gmail.com> 

jsancio committed:
MINOR; Fix partition change record noop check (#12073)

When LeaderRecoveryState was added to the PartitionChangeRecord, the
check for being a noop was not updated. This commit fixes that and
improves the associated test to avoid this oversight in the future.

Reviewers: Colin Patrick McCabe <cmccabe@apache.org>

showuon committed:
MINOR: revert back to 60s session timeout for static membership test (#11881)

Reviewers: Guozhang Wang <wangguoz@gmail.com>

cmccabe committed:
KAFKA-13841: Fix a case where we were unable to place on fenced brokers in KRaft mode (#12075)

This PR fixes a case where we were unable to place on fenced brokers In KRaft mode. Specifically,
if we had a broker registration in the metadata log, but no associated heartbeat, previously the
HeartbeatManager would not track the fenced broker. This PR fixes this by adding this logic to the
metadata log replay path in ClusterControlManager.

Reviewers: David Arthur <mumrah@gmail.com>, dengziming <dengziming1993@gmail.com>

dajac committed:
MINOR: Rename `ZkVersion` to `PartitionEpoch` (#12071)

This patch does some initial cleanups in the context of KAFKA-13790. Mainly, it renames `ZkVersion` field to `PartitionEpoch` in the `LeaderAndIsrRequest`, the `LeaderAndIsr` and the `Partition`.

Reviewers: Jason Gustafson <jason@confluent.io>, dengziming <dengziming1993@gmail.com>

Hongten committed:
KAFKA-13852: Kafka Acl documentation bug for wildcard '*' (#12090)

The wildcard * in command without wrapped by single quote will be replaced into the file name under the current folder by bash. So we need to wrap with single quote. Update the doc and command option description.

Reviewers: dengziming <dengziming1993@gmail.com>, Luke Chen <showuon@gmail.com>

ruanliang-hualun committed:
KAFKA-13834: add test coverage for RecordAccumulatorTest (#12092)

Reviewers: Luke Chen <showuon@gmail.com>

RivenSun2 committed:
MINOR: Using enums for auto.offset.reset configuration (#12077)

Using enums instead of Strings for auto.offset.reset configuration

Reviewers: Divij Vaidya <divijvaidya13@gmail.com>, Luke Chen <showuon@gmail.com

hachikuji committed:
MINOR: Change `AlterPartition` validation order in `KafkaController` (#12032)

Currently we validate recovery state before checking leader epoch in `KafkaController`. It seems more intuitive to validate leader epoch first since the leader might be working with stale state, which is what we do in KRaft. This patch fixes this and adds a couple additional validations to make the behavior consistent. 

Reviewers: José Armando García Sancio <jsancio@users.noreply.github.com>

dajac committed:
MINOR: Refactor `kafka.cluster.Replica` (#12081)

This patch refactors kafka.cluster.Replica, it usages and tests. This is part of the work in KAFKA-13790.

Reviewers: Jason Gustafson <jason@confluent.io>

ahuang98 committed:
MINOR; enable KRaft in ConfigCommandIntegrationTest (#11732)

Adding KRaft and ZK params to ConfigCommandIntegrationTest wherever appropriate.

Reviewers: Kvicii <42023367+Kvicii@users.noreply.github.com>, dengziming <dengziming1993@gmail.com>, José Armando García Sancio <jsancio@users.noreply.github.com>

philipnee committed:
KAFKA-12841: Remove an additional call of onAcknowledgement (#12064)

The bug was introduced in #11689 that an additional onAcknowledgement was made using the InterceptorCallback class. This is undesirable since onSendError will attempt to call onAcknowledgement once more.

Reviewers: Jun Rao <junrao@gmail.com>

hachikuji committed:
MINOR: Rename `AlterIsrManager` to `AlterPartitionManager` (#12089)

Since we have changed the `AlterIsr` API to `AlterPartition`, it makes sense to rename `AlterIsrManager` as well and some of the associated classes.

Reviewers: dengziming <dengziming1993@gmail.com>, David Jacot <djacot@confluent.io>

mt206 committed:
MINOR: fix html generation syntax errors (#12094)

The html document generation has some errors in it, specifically related to protocols. The two issues identified and resolved are:

* Missing </tbody> closing tags added
* Invalid usage of a <p> tag as a wrapper element for <table> elements. Changed the <p> tag to be a <div>.

Tested by running ./gradlew siteDocsTar and observing that the output was properly formed.

Reviewers: Guozhang Wang <wangguoz@gmail.com>

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java:
```diff
@@ -184,6 +247,78 @@ public void process(final Record<KIn, VIn> record) {
+                }
+            }
+
+            final long startNs = time.nanoseconds();
```
If we do intend to use nanoseconds instead of milliseconds, then we should name the metrics name as "...-latency-ns" and also in the description to emphasize it is measured in nanos, since by default all latency are measured in millis across AK package unless otherwise explicitly named / described.

Personally I think it's sufficient to measure in milis. WDYT @lihaosky @mjsax ?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/ProcessorNodeMetrics.java:
```diff
@@ -62,6 +67,17 @@ private ProcessorNodeMetrics() {}
+    private static final String EMITTED_RECORDS_RATE_DESCRIPTION =
+        RATE_DESCRIPTION_PREFIX + EMITTED_RECORDS_DESCRIPTION + RATE_DESCRIPTION_SUFFIX;
+
+    private static final String EMIT_FINAL_LATENCY = "window-aggregate-final-emit" + LATENCY_SUFFIX;
```
Please see my other comment --- if we measure in nanos we'd need to explicitly add that in the name and in the description.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/ProcessorNodeMetrics.java:
```diff
@@ -62,6 +67,17 @@ private ProcessorNodeMetrics() {}
+        RATE_DESCRIPTION_PREFIX + EMITTED_RECORDS_DESCRIPTION + RATE_DESCRIPTION_SUFFIX;
+
+    private static final String EMIT_FINAL_LATENCY = "window-aggregate-final-emit" + LATENCY_SUFFIX;
+    private static final String EMIT_FINAL_DESCRIPTION = "calls to emit final";
```
We can replace with `EMITTED_RECORDS + LATENCY_SUFFIX`.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java:
```diff
@@ -0,0 +1,1246 @@
+
+        builder.stream(TOPIC,
+            Consumed.with(Serdes.String(), Serdes.String()))
+            .transform(() -> new Transformer<String, String, KeyValue<String, String>>() {
```
I think all these transforms can be replaced as process since there's no resulted stream to continue.

guozhangwang replied:
EDIT: never mind, I saw that you've done so in the other PR...

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java:
```diff
@@ -0,0 +1,1246 @@
+    public void shouldSkipNonExistBaseKeyInCache() {
+        cachingStore.put(bytesKey("aa"), bytesValue("0002"), 0);
+
+        final SegmentedCacheFunction baseCacheFunction = new SegmentedCacheFunction(new TimeFirstWindowKeySchema(), SEGMENT_INTERVAL);
```
This is never used.

guozhangwang committed:
incorporate comments

guozhangwang reviewed (commented):
I incorporated the latest comments and pushed a new commit but it seems I messed github with other commits from trunk as well..

guozhangwang commented:
Sorry about that @lihaosky , I just created a copy PR from the branch in order to merge sooner: https://github.com/apache/kafka/pull/12100

@mjsax could you take a look at that PR.
